File: 2336-smallest-number-in-infinite-set/2336-smallest-number-in-infinite-set.py
class SmallestInfiniteSet:
    def __init__(self):
        self.heap = list()
        self.isIn = set()
        self.smallest = 1


    def popSmallest(self) -> int:
        if self.heap and self.heap[0] < self.smallest:
            target = heapq.heappop(self.heap)
            self.isIn.discard(target)
            return target
        self.smallest += 1
        return self.smallest - 1


    def addBack(self, num: int) -> None:
        if self.smallest > num and num not in self.isIn:
            heapq.heappush(self.heap, num)
            self.isIn.add(num)

File: 2336-smallest-number-in-infinite-set/2336-smallest-number-in-infinite-set.java
class SmallestInfiniteSet {
    PriorityQueue<Integer> heap;
    Set<Integer> set;
    int smallest;

    public SmallestInfiniteSet() {
        heap = new PriorityQueue<>();
        set = new HashSet<>();
        smallest = 1;
    }
    
    public int popSmallest() {
        if (!heap.isEmpty() && heap.peek() < smallest) {
            int target = heap.poll();
            set.remove(target);
            return target;
        }
        smallest++;
        return smallest - 1;
    }
    
    public void addBack(int num) {
        if (smallest > num && !set.contains(num)) {
            heap.add(num);
            set.add(num);
        }
    }
}